src/main/java/loci/common/S3Handle.java

```diff
@@ -0,0 +1,448 @@
+   * @throws IOException if there was an error connecting to the server
+   */
+  protected void connect() throws IOException {
+    try {
+      s3Client = new MinioClient(server, port, accessKey, secretKey);
+    }
+    catch (MinioException e) {
+      throw new IOException(String.format(
+              "Failed to connect: %s", this), e);
+    }
+    LOGGER.trace("connected:{}", this);
+  }
+
+  /**
+   * Check bucket or object exists
+   * @throws IOException if unable to get the object
+   * @throws MinioException if unable to get the object
+   * @throws InvalidKeyException if unable to get the object
+   * @throws NoSuchAlgorithmException if unable to get the object
+   * @throws XmlPullParserException if unable to get the object
+   */
+  protected void initialize() throws
+      IOException,
+      MinioException,
+      InvalidKeyException,
+      NoSuchAlgorithmException,
+      XmlPullParserException {
+    if (path == null) {
+      isBucket = s3Client.bucketExists(bucket);
+    }
+    else {
+      isBucket = false;
+      stat = s3Client.statObject(bucket, path);
+      resetStream();
+    }
+  }
+
+  public String getServer() {
+    return server;
+  }
+
+  public int getPort() {
+    return port;
+  }
+
+  public String getBucket() {
+    return bucket;
+  }
+
+  public String getPath() {
+    return path;
+  }
+
+  /**
+   * Download an S3 object to a file system cache if it doesn't already exist
+   *
+   * @param url the full URL to the S3 resource
+   * @param s custom settings object
+   * @return File path to the cached object
+   * @throws IOException if there is an error during reading or writing
+   * @throws HandleException if no destination for the cache is provided
+   */
+  public static String cacheObject(String url, Settings s) throws
+      IOException,
+      HandleException {
+    String cacheroot = s.getRemoteCacheRootDir();
+    if (cacheroot == null) {
+      throw new HandleException("Remote cache root dir is not set");
+    }
+    S3Handle s3 = new S3Handle(url, true, s);
+    // TODO: Need to ensure this path is safe. Is there a Java method to check?
+    String cacheobj = s3.getCacheKey();
+    // Hopefully creates a cross-platform path
+    Path cachepath = Paths.get(cacheroot, cacheobj);
+
+    if (Files.exists(cachepath)) {
+      LOGGER.debug("Found existing cache for {} at {}", s3, cachepath);
+    }
+    else {
+      LOGGER.debug("Caching {} to {}", s3, cachepath);
+      s3.downloadObject(cachepath);
+      LOGGER.debug("Downloaded {}", cachepath);
+    }
+    return cachepath.toString();
+  }
+
+  public String getCacheKey(){
+    String cachekey =
+      getServer().replace("://", "/") + "/" +
+      getPort() + "/" +
+      getBucket() + "/" +
+      getPath();
+    return cachekey;
+  }
+
+  protected void downloadObject(Path destination) throws HandleException, IOException {
+    LOGGER.trace("destination:{}", destination);
+    if (this.stat == null || this.objectNotFound != null) {
```

Comment:
In this and other places where ```objectNotFound``` is checked, does it ever make sense to attempt a call to ```initalize()``` again (in case of a network blip or something)?

Reply:
This might be something to look at in the future, but for the initial implementation I'd like to keep things simple and avoid automatic retries.

I've changed the constructor slightly so that if `this.initialize()` throws `IOException` it is no longer caught. Other exceptions are caught and held, these are mostly protocol/API errors so unlikely to be fixed by a retry: http://minio.github.io/minio-java/io/minio/errors/ErrorResponseException.html

I think the ideal situation would be to not catch any Exceptions at all and convert everything that throws in the constructor into `IOException`, but this causes problems in the `Location` class where it's assumed a handle can be created to a non-existent object.